fronted.go

```diff
@@ -139,42 +212,82 @@ func Vet(m *Masquerade, pool *x509.CertPool, testURL string) bool {
+// findWorkingFronts finds working domain fronts by testing them using a worker pool. Speed
+// is of the essence here, as without working fronts, users will
 // be unable to fetch proxy configurations, particularly in the case of a first time
 // user who does not have proxies cached on disk.
-func (f *fronted) findWorkingMasquerades(listener func(f *fronted)) {
-	// vet masquerades in batches
-	const batchSize int = 40
-	var successful atomic.Uint32
-
-	// We loop through all of them until we have 4 successful ones.
-	for i := 0; i < len(f.masquerades) && successful.Load() < 4; i += batchSize {
-		f.vetBatch(i, batchSize, &successful, listener)
+func (f *fronted) findWorkingFronts() {
+	// Keep looping through all fronts making sure we have working ones.
+	for {
+		// Continually loop through the fronts until we have 4 working ones.
+		// This is important, for example, when the user goes offline and all fronts start failing.
+		// We want to just keep trying in that case so that we find working fronts as soon as they
+		// come back online.
+		if !f.hasEnoughWorkingFronts() {
+			// Note that trying all fronts takes awhile, as it only completes when we either
+			// have enough working fronts, or we've tried all of them.
+			log.Debug("findWorkingFronts::Trying all fronts")
+			f.tryAllFronts()
+			log.Debug("findWorkingFronts::Tried all fronts")
+		} else {
+			log.Debug("findWorkingFronts::Enough working fronts...sleeping")
+			select {
+			case <-f.stopCh:
+				log.Debug("findWorkingFronts::Stopping parallel dialing")
+				return
+			case <-time.After(time.Duration(rand.IntN(12000)) * time.Millisecond):
+				// Run again after a random time between 0 and 12 seconds
+			}
+		}
 	}
 }
 
-func (f *fronted) vetBatch(start, batchSize int, successful *atomic.Uint32, listener func(f *fronted)) {
-	log.Debugf("Vetting masquerade batch %d-%d", start, start+batchSize)
-	var wg sync.WaitGroup
-	masqueradeSize := len(f.masquerades)
-	for i := start; i < start+batchSize && i < masqueradeSize; i++ {
-		wg.Add(1)
-		go func(m MasqueradeInterface) {
-			defer wg.Done()
-			if f.vetMasquerade(m) {
-				successful.Add(1)
-				if listener != nil {
-					go listener(f)
-				}
+func (f *fronted) tryAllFronts() {
+	// Vet fronts using a worker pool of 40 goroutines.
+	pool := pond.NewPool(40)
+
+	// Submit all fronts to the worker pool.
+	for i := 0; i < f.frontSize(); i++ {
+		i := i
+		m := f.frontAt(i)
+		pool.Submit(func() {
+			log.Debugf("Running task #%d with front %v", i, m.getIpAddress())
+			if f.hasEnoughWorkingFronts() {
+				// We have enough working fronts, so no need to continue.
+				log.Debug("Enough working fronts...ignoring task")
+				return
 			}
-		}(f.masquerades[i])
+			working := f.vetFront(m)
+			if working {
+				f.connectingFronts.onConnected(m)
+			} else {
+				m.markFailed()
+			}
+		})
 	}
-	wg.Wait()
+
+	// Stop the pool and wait for all submitted tasks to complete
+	pool.StopAndWait()
+}
```

Comment:
We should probably still stop vetting if `Close` is called. Otherwise, this looks great!